### PR TITLE
Add workspace-loader flow

### PR DESCRIPTION
**Pull request by olexii4**
Covers Workspace Loader section of https://github.com/eclipse/che/issues/14126

Signed-off-by: Oleksii <oorel@redhat.com>

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+
+```bash
+# [ws-loader-dev]
+$ yarn build
+```
+
+#### Step 4: Code ...
+
+#### Step 5: Run unit test (optional)
+In this step, we will run the Workspace Lloader unit tests:
```
Comment (mmorhun) [inline CONTRIBUTING.md]: Typo in `Lloader`

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+In this section, we show how to setup a Che 7 environment to work on the che workspace loader, and how to use it.
+For the whole workflows, we will need a workspace with such containers:
+
+- Workspace Loader Dev container (a.k.a ws-loader-dev) is a all in one container for running commands such as build, test or start the workspace loader server.
+
+All containers have `/projects` folder mounted, which is shared among them.
+
+Developer workflow:
+
+1. Start the workspace with the devfile, it is cloning Che repo.
```
Comment (mmorhun) [inline CONTRIBUTING.md]: I think current title is too detailed for contents. I would name it `Create and start the workspace` and explain about cloning into actual section.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
 
+### Workspace Loader
+Workspace loader source code is located in [<this.repo>/workspace-loader](./workspace-loader/) folder.
+Here is the developer workflow if you want to contribute to it:
+
+#### Devfile for workspace loader development
+
+The devfile: [<this.repo>/workspace-loader/devfile.yaml](./workspace-loader/devfile.yaml)
+
+In this section, we show how to setup a Che 7 environment to work on the che workspace loader, and how to use it.
```
Comment (mmorhun) [inline CONTRIBUTING.md]: Please use capital letter for Che

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+For the whole workflows, we will need a workspace with such containers:
```
Comment (mmorhun) [inline CONTRIBUTING.md]: If we actually need only one container, maybe rewrite this one item list to just a container... But up to you.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+
+- Workspace Loader Dev container (a.k.a ws-loader-dev) is a all in one container for running commands such as build, test or start the workspace loader server.
```
Comment (mmorhun) [inline CONTRIBUTING.md]: Is `ws-loader-dev` a common name in ws-loader development?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+
+All containers have `/projects` folder mounted, which is shared among them.
+
+Developer workflow:
+
+1. Start the workspace with the devfile, it is cloning Che repo.
+2. Install dependencies
+3. Build
+4. Code ...
```
Comment (mmorhun) [inline CONTRIBUTING.md]: `Make changes`?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+5. Run unit test
```
Comment (mmorhun) [inline CONTRIBUTING.md]: Typo: `tests`

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+6. Start workspace loader server and preview
```
Comment (mmorhun) [inline CONTRIBUTING.md]: To me `and test` sounds better.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+
+#### Step 1: Start the workspace with the devfile, it is cloning Che repo.
+In this section we are going to start a new workspace to work on workspace loader. The new workspace will clone the `che` project. It will also setup the containers and commands in the `My workspace` view. We will use these commands in the next steps.
```
Comment (mmorhun) [inline CONTRIBUTING.md]: Please use `Eclipse Che` instead of `che`

Comment (mmorhun) [inline CONTRIBUTING.md]: `It will also setup the containers and commands` -> `Containers and commands could be found in the My Workspace panel`

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+
+or as a factory `https://<CheInstance>/f?url=https://raw.githubusercontent.com/eclipse/che/master/workspace-loader/devfile.yaml`
+
+At workspace start, Che will clone Che source code (including the workspace-loader)
+
+#### Step 2: Install dependencies
+
+In this section we are going to install workspace loader dependencies.
+
+You can use the che command `[workspace loader] install dependencies` (command pallette > Run task > … or containers view)
```
Comment (mmorhun) [inline CONTRIBUTING.md]: I think just `use the command` will be enough.

Comment (mmorhun) [inline CONTRIBUTING.md]: `command pallette > Run task > … or containers view` -> `Terminal -> `Run Tusk` -> `TASK NAME HERE` or use `My Workspace` panel and just click on the command.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+```bash
+# [ws-loader-dev]
+$ yarn
+```
+
+#### Step 3: Build
+
+In this section we are going to build the workspace loader project.
+
+You can use the che command `[workspace loader] run build` (command pallette > Run task > … or containers view)
```
Comment (mmorhun) [inline CONTRIBUTING.md]: The same as in Step 2

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+Basically, this command will run
+
+```bash
+# [ws-loader-dev]
+$ yarn build
+```
+
+#### Step 4: Code ...
+
```
Comment (mmorhun) [inline CONTRIBUTING.md]: I would write a trivial sentence like `Now you can make changes in Workspace Loader`

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -113,6 +111,89 @@ You can use the che command `dashboard_dev_server` (command pallette > Run task
...
+You can use the che command `[workspace loader] install dependencies` (command pallette > Run task > … or containers view)
+Basically, this command will run
+
+```bash
+# [ws-loader-dev]
+$ yarn
+```
+
+#### Step 3: Build
```
Comment (mmorhun) [inline CONTRIBUTING.md]: Please explain in the documentation why we need build before actual changes

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -104,15 +110,106 @@ Basically, this command will run
...
+
+You can use the che command `[workspace loader] run build` (command pallette > Run task > … or containers view)
+Basically, this command will run
+
+```bash
+# [ws-loader-dev]
+$ yarn build
+```
+
+#### Step 4: Code ...
```
Comment (AndrienkoAleksandr) [inline CONTRIBUTING.md]: Maybe apply todo... or coming soon? It's not clear why this section is empty.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -104,15 +110,106 @@ Basically, this command will run
...
+### Workspace Loader
+
+Workspace loader source code is located in [<this.repo>/workspace-loader](./workspace-loader/) folder.
+Here is the developer workflow if you want to contribute to it:
+
+#### Devfile for workspace loader development
+
+The devfile: [<this.repo>/workspace-loader/devfile.yaml](./workspace-loader/devfile.yaml)
+
+This section shows you how to setup a Che 7 environment to work on the che workspace loader, and how to use it.
```
Comment (mmorhun) [inline CONTRIBUTING.md]: Please use `Che`

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -104,15 +110,106 @@ Basically, this command will run
...
+
+You can use the che command `[workspace loader] run tests` (command pallette > Run task > … or containers view)
+Basically, this command will run
+
+```bash
+# [ws-loader-dev]
+$ yarn test
+```
+
+#### Step 6: Start  workspace loader  server and preview
```
Comment (mmorhun) [inline CONTRIBUTING.md]: Extra spaces: `Start  workspace` and `loader  server`

**Review by mmorhun — COMMENTED** [review]
Giving some suggestions...

File: CONTRIBUTING.md (hunk 16 above)
Comment (AndrienkoAleksandr) [reply CONTRIBUTING.md]: And `Start  workspace` too.

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -104,15 +110,106 @@ Basically, this command will run
...
+Here is the developer workflow if you want to contribute to it:
+
+#### Devfile for workspace loader development
+
+The devfile: [<this.repo>/workspace-loader/devfile.yaml](./workspace-loader/devfile.yaml)
+
+This section shows you how to setup a Che 7 environment to work on the che workspace loader, and how to use it.
+For the whole workflows, we will need a workspace with such containers:
+
+- Workspace Loader Dev container (a.k.a ws-loader-dev) is a all in one container for running commands such as build, test or start the workspace loader server.
```
Comment (AndrienkoAleksandr) [inline CONTRIBUTING.md]: Maybe `is a all in one container` => `is all in one container` ?

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -104,15 +110,106 @@ Basically, this command will run
...
+
+All containers have `/projects` folder mounted, which is shared among them.
```
Comment (AndrienkoAleksandr) [inline CONTRIBUTING.md]: > All containers have `/projects` folder mounted, which is shared among them.

Actually not all containers have `/projects` volume. Only containers marked in the devfile with attribute:  `mountSources: true`

File: CONTRIBUTING.md (hunk 13 above)
Comment (olexii4) [reply CONTRIBUTING.md]: Just to be sure about the workspace environment and the build process with last changes

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -12,39 +12,45 @@
...
 ```bash
-$ chectl workspace:start -f devfiles/che-theia-all.devfile.yaml
+chectl workspace:start -f devfiles/che-theia-all.devfile.yaml
 ```
+
 or
+
 ```bash
-$ chectl workspace:start -f https://raw.githubusercontent.com/eclipse/che-theia/master/devfiles/che-theia-all.devfile.yaml
+chectl workspace:start -f https://raw.githubusercontent.com/eclipse/che-theia/master/devfiles/che-theia-all.devfile.yaml
```
Comment (AndrienkoAleksandr) [inline CONTRIBUTING.md]: Why Did you remove `$` in a few places? For most users this character associated with command line interface.

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -104,15 +110,106 @@ Basically, this command will run
...
+
+You can use the che command `[workspace loader] run build` (command pallette > Run task > … or containers view)
+Basically, this command will run
+
+```bash
+# [ws-loader-dev]
+$ yarn build
+```
+
+#### Step 4: Code ...
```
Comment (AndrienkoAleksandr) [inline CONTRIBUTING.md]: As for me empty section looks strange. It would be nice to apply at least one sentences with explanation. But up to you.

File: CONTRIBUTING.md (hunk 5 above)
Comment (olexii4) [reply CONTRIBUTING.md]: Yes It Is
https://github.com/eclipse/che/blob/master/workspace-loader/devfile.yaml#L19

File: CONTRIBUTING.md (hunk 15 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 4 above)
Comment (olexii4) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md (hunk 17 above)
Comment (olexii4) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md (hunk 18 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 1 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 2 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 3 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 4 above)
Comment (olexii4) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md (hunk 5 above)
Comment (olexii4) [reply CONTRIBUTING.md]: Yes it is
https://github.com/eclipse/che/blob/master/workspace-loader/devfile.yaml#L19

File: CONTRIBUTING.md (hunk 6 above)
Comment (olexii4) [reply CONTRIBUTING.md]: removed

File: CONTRIBUTING.md (hunk 7 above)
Comment (olexii4) [reply CONTRIBUTING.md]: removed

File: CONTRIBUTING.md (hunk 9 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 10 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 17 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 19 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 20 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 18 above)
Comment (olexii4) [reply CONTRIBUTING.md]: removed

File: CONTRIBUTING.md (hunk 16 above)
Comment (olexii4) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md (hunk 15 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 12 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 11 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 10 above)
Comment (olexii4) [reply CONTRIBUTING.md]: done

File: CONTRIBUTING.md (hunk 8 above)
Comment (olexii4) [reply CONTRIBUTING.md]: I do not think so because we already have the test command.

**Review by mmorhun — APPROVED** [review]
Good job!

**Review by akurinnoy — APPROVED** [review]

**Review by AndrienkoAleksandr — APPROVED** [review]

**Review by vparfonov — APPROVED** [review]
